base/compiler/ssair/ir.jl

```diff
@@ -754,23 +791,137 @@ function renumber_ssa2!(@nospecialize(stmt), ssanums::Vector{Any}, used_ssa::Vec
+    deleteat!(preds, findfirst(x->x === compact.bb_rename[from], preds)::Int)
+    deleteat!(succs, findfirst(x->x === compact.bb_rename[to], succs)::Int)
+    # Check if the block is now dead
+    if length(preds) == 0
+        for succ in copy(compact.result_bbs[compact.bb_rename[to]].succs)
+            kill_edge!(compact, active_bb, to, findfirst(x->x === succ, compact.bb_rename))
+        end
+        if to < active_bb
+            # Kill all statements in the block
+            stmts = compact.result_bbs[compact.bb_rename[to]].stmts
+            for stmt in stmts
+                compact.result[stmt] = nothing
+            end
+            compact.result[last(stmts)] = ReturnNode()
+        end
+    else
+        # We need to remove this edge from any phi nodes
+        if to < active_bb
+            idx = first(compact.result_bbs[compact.bb_rename[to]].stmts)
+            while idx < length(compact.result) &&
+                    (stmt = compact.result[idx]; isa(stmt, PhiNode) || stmt === nothing)
+                stmt === nothing && continue
+                i = findfirst(x-> x === compact.bb_rename[from], stmt.edges)
+                if i !== nothing
+                    deleteat!(stmt.edges, i)
+                    deleteat!(stmt.values, i)
+                end
+                idx += 1
+            end
+        else
+            idx = first(compact.ir.cfg.blocks[to].stmts)
+            for stmt in CompactPeekIterator(compact, idx)
+                stmt === nothing && continue
+                isa(stmt, PhiNode) || break
+                i = findfirst(x-> x === from, stmt.edges)
+                if i !== nothing
+                    deleteat!(stmt.edges, i)
+                    deleteat!(stmt.values, i)
+                end
+            end
+        end
+    end
+end
+
+function process_node!(compact::IncrementalCompact, result::Vector{Any},
+        result_idx::Int, ssa_rename::Vector{Any},
         late_fixup::Vector{Int}, used_ssas::Vector{Int}, @nospecialize(stmt),
-        idx::Int, processed_idx::Int, do_rename_ssa::Bool)
+        idx::Int, processed_idx::Int, active_bb::Int, do_rename_ssa::Bool)
     ssa_rename[idx] = SSAValue(result_idx)
     if stmt === nothing
         ssa_rename[idx] = stmt
     elseif isa(stmt, OldSSAValue)
         ssa_rename[idx] = ssa_rename[stmt.id]
-    elseif isa(stmt, GotoNode) || isa(stmt, GlobalRef)
+    elseif isa(stmt, GotoNode) && compact.allow_cfg_transforms
+        result[result_idx] = GotoNode(compact.bb_rename[stmt.label])
+        result_idx += 1
+    elseif isa(stmt, GlobalRef) || isa(stmt, GotoNode)
+        result[result_idx] = stmt
+        result_idx += 1
+    elseif isa(stmt, GotoIfNot) && compact.allow_cfg_transforms
+        stmt = renumber_ssa2!(stmt, ssa_rename, used_ssas, late_fixup, result_idx, do_rename_ssa)::GotoIfNot
+        @assert stmt.dest !== active_bb+1
```

Comment:
Double edges were disallowed in a prior PR.

Reply:
Do we believe it's impossible for this PR to attempt to introduce this? For example after eliminating the contents of a block, the GotoIfNot should also get removed:
```
if condition
nothing
end
```

Reply:
I'm not sure what you're saying here. No reachable block should ever get removed, so the only way we could introduce this is if we were to combine blocks, which we don't anywhere here. Am I missing something? In either case, I can remove this assertion, because it doesn't make too much sense here. It is still checked in the verifier though.